File: gaming_deals_bot/epic.py
import logging
from dataclasses import dataclass
from datetime import datetime, timezone

import httpx

logger = logging.getLogger(__name__)

FREE_GAMES_URL = (
    "https://store-site-backend-static.ak.epicgames.com/freeGamesPromotions"
)


@dataclass
class EpicFreeGame:
    title: str
    game_id: str
    description: str
    end_date: str | None  # ISO datetime when the offer expires
    url: str

    @property
    def dedup_id(self) -> str:
        return f"epic-{self.game_id}"


def _parse_promotions(
    elements: list[dict],
) -> tuple[list[EpicFreeGame], list[EpicFreeGame]]:
    """Parse Epic promotions into current free games and upcoming free games."""
    current: list[EpicFreeGame] = []
    upcoming: list[EpicFreeGame] = []
    now = datetime.now(timezone.utc)

    for elem in elements:
        title = elem.get("title", "Unknown")
        game_id = elem.get("id", "")
        description = elem.get("description", "")

        # Build the store URL from the slug or product slug
        slug = (
            elem.get("productSlug")
            or elem.get("urlSlug")
            or elem.get("catalogNs", {}).get("mappings", [{}])[0].get("pageSlug", "")
        )
        url = f"https://store.epicgames.com/en-US/p/{slug}" if slug else ""

        # Check current promotional offers
        promotions = elem.get("promotions")
        if not promotions:
            continue

        # Current offers
        for offer_group in promotions.get("promotionalOffers", []):
            for offer in offer_group.get("promotionalOffers", []):
                discount = offer.get("discountSetting", {})
                if discount.get("discountPercentage", 100) == 0:
                    end_date = offer.get("endDate")
                    # Verify the offer is currently active
                    start = offer.get("startDate")
                    if start:
                        start_dt = datetime.fromisoformat(
                            start.replace("Z", "+00:00")
                        )
                        if start_dt > now:
                            continue
                    if end_date:
                        end_dt = datetime.fromisoformat(
                            end_date.replace("Z", "+00:00")
                        )
                        if end_dt < now:
                            continue

                    current.append(
                        EpicFreeGame(
                            title=title,
                            game_id=game_id,
                            description=description,
                            end_date=end_date,
                            url=url,
                        )
                    )

        # Upcoming offers
        for offer_group in promotions.get("upcomingPromotionalOffers", []):
            for offer in offer_group.get("promotionalOffers", []):
                discount = offer.get("discountSetting", {})
                if discount.get("discountPercentage", 100) == 0:
                    end_date = offer.get("endDate")
                    upcoming.append(
                        EpicFreeGame(
                            title=title,
                            game_id=game_id,
                            description=description,
                            end_date=end_date,
                            url=url,
                        )
                    )

    return current, upcoming


async def fetch_free_games(
    client: httpx.AsyncClient,
) -> tuple[list[EpicFreeGame], list[EpicFreeGame]]:
    """Fetch current and upcoming free games from Epic Games Store.

    Returns (current_free_games, upcoming_free_games).
    """
    try:
        resp = await client.get(FREE_GAMES_URL, params={"locale": "en-US"})
        resp.raise_for_status()
        data = resp.json()
    except (httpx.HTTPError, ValueError) as exc:
        logger.error("Epic Games Store API error: %s", exc)
        return [], []

    elements = (
        data.get("data", {})
        .get("Catalog", {})
        .get("searchStore", {})
        .get("elements", [])
    )

    if not elements:
        logger.warning("No elements found in Epic free games response")
        return [], []

    current, upcoming = _parse_promotions(elements)
    logger.info(
        "Epic: %d current free games, %d upcoming",
        len(current),
        len(upcoming),
    )
    return current, upcoming

File: internal/deals/epic.go
package deals

import (
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"time"
)

type EpicFreeGame struct {
	ID       string
	Title    string
	Desc     string
	URL      string
	EndDate  *time.Time
	Upcoming bool
	DedupID  string
}

type epicResponse struct {
	Data struct {
		Catalog struct {
			SearchStore struct {
				Elements []epicElement `json:"elements"`
			} `json:"searchStore"`
		} `json:"Catalog"`
	} `json:"data"`
}

type epicElement struct {
	ID          string `json:"id"`
	Title       string `json:"title"`
	Description string `json:"description"`
	ProductSlug string `json:"productSlug"`
	URLSlug     string `json:"urlSlug"`
	CatalogNs   struct {
		Mappings []struct {
			PageSlug string `json:"pageSlug"`
		} `json:"mappings"`
	} `json:"catalogNs"`
	Promotions *struct {
		PromotionalOffers         []epicOfferGroup `json:"promotionalOffers"`
		UpcomingPromotionalOffers []epicOfferGroup `json:"upcomingPromotionalOffers"`
	} `json:"promotions"`
}

type epicOfferGroup struct {
	PromotionalOffers []epicOffer `json:"promotionalOffers"`
}

type epicOffer struct {
	StartDate       string `json:"startDate"`
	EndDate         string `json:"endDate"`
	DiscountSetting struct {
		DiscountPercentage int `json:"discountPercentage"`
	} `json:"discountSetting"`
}

// FetchEpicFreeGames fetches current and upcoming free games from Epic Games Store.
func FetchEpicFreeGames() ([]EpicFreeGame, error) {
	resp, err := http.Get("https://store-site-backend-static.ak.epicgames.com/freeGamesPromotions?locale=en-US")
	if err != nil {
		return nil, fmt.Errorf("epic request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("epic returned status %d", resp.StatusCode)
	}

	var result epicResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("epic decode failed: %w", err)
	}

	now := time.Now()
	var games []EpicFreeGame

	for _, elem := range result.Data.Catalog.SearchStore.Elements {
		if elem.Promotions == nil {
			continue
		}

		slug := elem.ProductSlug
		if slug == "" {
			slug = elem.URLSlug
		}
		if slug == "" && len(elem.CatalogNs.Mappings) > 0 {
			slug = elem.CatalogNs.Mappings[0].PageSlug
		}
		if slug == "" {
			slog.Warn("epic: skipping game with no slug", "title", elem.Title, "id", elem.ID)
			continue
		}
		storeURL := fmt.Sprintf("https://store.epicgames.com/en-US/p/%s", slug)

		// Check current free offers
		for _, group := range elem.Promotions.PromotionalOffers {
			for _, offer := range group.PromotionalOffers {
				if offer.DiscountSetting.DiscountPercentage != 0 {
					continue
				}
				start, errS := time.Parse(time.RFC3339, offer.StartDate)
				end, errE := time.Parse(time.RFC3339, offer.EndDate)
				if errS != nil || errE != nil {
					slog.Warn("epic: failed to parse offer dates", "title", elem.Title, "start", offer.StartDate, "end", offer.EndDate)
					continue
				}

				if start.After(now) || end.Before(now) {
					continue
				}

				game := EpicFreeGame{
					ID:       elem.ID,
					Title:    elem.Title,
					Desc:     elem.Description,
					URL:      storeURL,
					Upcoming: false,
					DedupID:  fmt.Sprintf("epic-current-%s", elem.ID),
				}
				if !end.IsZero() {
					game.EndDate = &end
				}
				games = append(games, game)
			}
		}

		// Check upcoming free offers
		for _, group := range elem.Promotions.UpcomingPromotionalOffers {
			for _, offer := range group.PromotionalOffers {
				if offer.DiscountSetting.DiscountPercentage != 0 {
					continue
				}

				game := EpicFreeGame{
					ID:       elem.ID,
					Title:    elem.Title,
					Desc:     elem.Description,
					URL:      storeURL,
					Upcoming: true,
					DedupID:  fmt.Sprintf("epic-upcoming-%s", elem.ID),
				}
				if offer.EndDate != "" {
					if end, err := time.Parse(time.RFC3339, offer.EndDate); err == nil {
						game.EndDate = &end
					}
				}
				games = append(games, game)
			}
		}
	}

	return games, nil
}
